parser/src/grammar.lalrpop

```diff
@@ -0,0 +1,144 @@
+use crate::{
+    ast::{
+        Boundary, BoundaryConstraints, Constraint, Expr, AirDef,
+        Identifier, Source, SourceSection, TraceCols, TraceColsGrp,
+        TraceColsGrpType, TransitionConstraints
+    }, error::{Error, ParseError::InvalidInt}, lexer::Tok
+};
+use std::str::FromStr;
+use lalrpop_util::ParseError;
+
+grammar;
+
+// AST NODE
+// ------------------------------------------------------------------------------------------------
+
+pub Source: Source = {
+    SourceSection* => Source(<>)
+}
+
+SourceSection: SourceSection = {
+    AirDef => SourceSection::AirDef(<>),
+    BoundaryConstraints => SourceSection::BoundaryConstraints(<>),
+    TraceCols => SourceSection::TraceCols(<>),
+    TransitionConstraints => SourceSection::TransitionConstraints(<>),
+}
+
+// FILE DEF
+// ------------------------------------------------------------------------------------------------
+
+AirDef: AirDef = {
+    "def" <f: Identifier> => AirDef { name: f }
+}
+
+// TRACE COLUMNS
+// ------------------------------------------------------------------------------------------------
+
+TraceCols: TraceCols = {
+    "trace_columns" ":" <cols: TraceColsGrp+> => TraceCols { cols }
+}
+
+TraceColsGrp: TraceColsGrp = {
+    "main" ":" "[" <e: CommaElems<Expr1>> "]" => TraceColsGrp {
+        cols_grp_type: TraceColsGrpType::MainTraceCols,
+        cols: e,
+    } 
+}
+
+// TRANSITION CONSTRAINTS
+// ------------------------------------------------------------------------------------------------
+
+TransitionConstraints: TransitionConstraints = {
+    "transition_constraints" ":" <transition_constraints: Constraint+> =>
+        TransitionConstraints { transition_constraints }
+}
+
+// BOUNDARY CONSTRAINTS
+// ------------------------------------------------------------------------------------------------
+
+BoundaryConstraints: BoundaryConstraints = {
+    "boundary_constraints" ":" <boundary_constraints: Constraint+> =>
+        BoundaryConstraints { boundary_constraints }
+}
+
+Boundary: Boundary = {
+    "first" => Boundary::First,
+    "last" => Boundary::Last
+}
```

Comment:
nit: I would put the boundary constraints section before the transition constraints section. It's not a big deal, but it's just useful to be consistent across all places, and that seems to be the logical order we follow elsewhere in Miden.